MONGOID-5151 Respect aliased fields in pluck/distinct by having Document.database_field_name recursively consider embedded docs

johnnyshields opened the pull request:
`Mongoid::Document.database_field_name` does not currently handle "." delimiter char. As an improvement, it should consider embedded doc `:store_as` and embedded field aliases.

As a positive side-effect of this PR, `distinct` and `pluck` (and maybe other methods) now magically work when using embedded field aliases.

This PR also adds `Mongoid::Document.aliased_relations` which is a map analogous to `aliased_fields`

This is a precursor to further improvements on "distinct" and "pluck" methods.

## TODO Tests:

* [ ] Test that embedded aliases work on `pluck` and `distinct` methods (this PR fixes that)
* [ ] Add tests for Document.aliased_relations

p-mongo commented:
Hi Johnny. Are you sure the ticket was created? I don't see any new tickets from you just now.

p-mongo commented:
I'm not sure you are aware, as of MongoDB 5.0 dots and dollars in field names are now explicitly supported. So, if a user wants to store in a field named `a.b`, this field name verbatim should generally be used.

Because of this a  change to how dots are treated needs to  be carefully considered. Can you add a user-facing example of what you are proposing the behavior to be? It doesn't need to be a test case, just something we can refer to for discussion purposes.

johnnyshields commented:
@p-mongo sure:

```ruby
Person.distinct('phones.number')
```

**This works currently**, if there are no aliases.

The periods in field names is going to royally mess up Mongoid, because we have `split(".")` literally everywhere as you know. We need a ticket to review that everywhere in code... Do we need to consider a new delimiters? Or do commands like distinct require an escape `\`

johnnyshields commented:
About JIRA... I have tried on two different browsers and I can't create tickets. Totally bizarre... it looks like the "Create Ticket" form saves, but the ticket isn't actually created.

johnnyshields commented:
@p-mongo please read this: https://docs.mongodb.com/manual/core/dot-dollar-considerations/#std-label-crud-concepts-dot-dollar-considerations


> MongoDB 5.0 adds improved support for field names that are dollar ($) prefixed or that contain periods (.)... In most cases data that has been stored using field names like these **is not directly accessible**. You need to use helper methods like $getField, $setField, and $literal in queries that access those fields.

So, since special operators are needed Mongoid / Ruby Driver basically won't work with such fields--certainly not today, and it's reasonable to say we shouldn't ever support them. In Mongoid we'd have to add a new atomic action in Mongoid specifically for setting `$` and `.` fields... I just don't think there's any real world use case for it. Really questionable why MongoDB added this feature in the first place...

johnnyshields commented:
🤦‍♂️ 

![image](https://user-images.githubusercontent.com/27655/130288803-bc9d71b9-10ff-4c9e-9288-86814f45e1de.png)

johnnyshields commented:
@p-mongo I've addressed concerns with `.` and `$` fields here: https://github.com/mongodb/mongoid/pull/5051

In summary, it is a non-issue for this PR.

johnnyshields commented:
@p-mongo can you please confirm your previous comment is no longer a concern?

p-mongo commented:
Hi @johnnyshields ,

I added integration tests and I was able to verify the `distinct` and `pluck` behavior changed as you described, for which I added documentation and release notes.

However for the query example I wasn't getting aliases expanded:

```
Person.where('phones.number' => '+12061112222')
```

Can you confirm whether the `where` call is also supposed to expand the aliases?

p-mongo commented on lib/mongoid/association/macros.rb:
```diff
@@ -12,9 +12,11 @@ module Macros
         self.embedded = false
         self.embedded_relations = BSON::Document.new
         self.relations = BSON::Document.new
+        self.aliased_relations = {}
```
If possible I would like new code to use the term "association" instead of "relation". I understand that the additions match the existing style but my preference nonetheless is to use correct terminology in new code. Do you have any objections to using "association" in new code? To clarify, none of the existing code should be changed at this time for BC reasons.

johnnyshields replied:
No objection, `aliased_associations` is fine.

johnnyshields commented:
@p-mongo 

> Can you confirm whether the `where` call is also supposed to expand the aliases?

Yes, `where` should expand the aliases. We should include that in this PR.

p-mongo commented on lib/mongoid/fields.rb:
```diff
@@ -252,18 +252,32 @@ def attribute_names
+          "#{segment}.#{relation.klass.database_field_name(remaining)}"
+        else
+          "#{segment}.#{remaining}"
+        end.freeze
```
What is the purpose of the freeze calls in this method please?

p-mongo commented:
This PR as far as I can tell is complete with respect to pluck and distinct, including documentation. Therefore my suggestion would be to merge it as is and tackle `where` in another PR.

For `where`, while I generally agree that aliased field expansion could be useful, I am wondering if the various mechanisms that Mongoid provides all would work with it (symbol operators, Mongoid methods and MQL provided as hashes).

johnnyshields commented:
Interpolated strings arent frozen by default.


On Fri, Sep 10, 2021 at 13:33 Oleg Pudeyev ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In lib/mongoid/fields.rb
> <https://github.com/mongodb/mongoid/pull/5047#discussion_r705888363>:
>
> >        def database_field_name(name)
> -        return nil unless name
> -        normalized = name.to_s
> -        aliased_fields[normalized] || normalized
> +        return nil unless name.present?
> +        key = name.to_s
> +        segment, remaining = key.split('.', 2)
> +        segment = aliased_fields[segment] || segment
> +        return segment.freeze unless remaining
> +
> +        relation = relations[aliased_relations[segment] || segment]
> +        if relation
> +          "#{segment}.#{relation.klass.database_field_name(remaining)}"
> +        else
> +          "#{segment}.#{remaining}"
> +        end.freeze
>
> What is the purpose of the freeze calls in this method please?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mongodb/mongoid/pull/5047#pullrequestreview-751039524>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAGYB4W2ZBVZN6MUUMAN5LUBGDA3ANCNFSM5CQVIRYQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>
-- 

Johnny Shields

Co-Founder & Chief Technology Officer

[image: TableCheck] <https://www.tablecheck.com/en/company/>
[image: Address] Dai 27 Chuo Bldg 4F, 2-14-5 Ginza, Chuo-ku, Tokyo, Japan
104-0061 <https://goo.gl/maps/G4PyDSZE8s32>
[image: Phone] +81 (0)80-3519-5647
[image: Web]  ***@***.***
https://www.tablecheck.com/en/company/

p-mongo commented:
Thanks for clarifying that. It's a bit hard to tell from the diff but it seems to me that previously the return value of the method was not frozen. Changing it to a frozen string would technically be an API change and we've had an issue in the past with users expecting unfrozen objects from our methods. Is freezing the strings as you originally wrote an optimization only or was it required for functionality?

johnnyshields commented:
Freezing not required. Let's go with your way.

p-mongo commented:
https://jira.mongodb.org/browse/MONGOID-5151

comandeo reviewed (approved):
@p-mongo What is the reason of adding commented out tests, Maybe we can use `pending`? I executed this file locally, and only test failed.

p-mongo commented:
I removed it and put it in the ticket for future use, if we end up needing it.

johnnyshields commented:
Thanks for the merge, I appreciate it!
